vcs/git/gogit/repo.go

```diff
@@ -0,0 +1,354 @@
+package git
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/shazow/go-git"
+	"github.com/shazow/go-vcs/vcs/gitcmd"
+	"golang.org/x/tools/godoc/vfs"
+	"sourcegraph.com/sourcegraph/go-vcs/vcs"
+	"sourcegraph.com/sqs/pbtypes"
+)
+
+func init() {
+	// Overwrite the git opener to return repositories that use the
+	// gogits native-go implementation.
+	vcs.RegisterOpener("git", func(dir string) (vcs.Repository, error) {
+		return Open(dir)
+	})
+}
+
+// Repository is a git VCS repository.
+type Repository struct {
+	// TODO: Do we need locking?
+	repo *git.Repository
+	// fallback is used for features that are not implemented here yet.
+	fallback *gitcmd.Repository
+}
+
+func (r *Repository) RepoDir() string {
+	return r.repo.Path
+}
+
+func (r *Repository) GitRootDir() string {
+	return r.RepoDir()
+}
+
+func (r *Repository) String() string {
+	return fmt.Sprintf("git (gogit) repo at %s", r.RepoDir())
+}
+
+func Open(dir string) (*Repository, error) {
+	if _, err := os.Stat(filepath.Join(dir, ".git")); !os.IsNotExist(err) {
+		// Append .git to path
+		dir = filepath.Join(dir, ".git")
+	}
+
+	repo, err := git.OpenRepository(dir)
+	if err != nil {
+		// FIXME: Wrap in vcs error?
+		return nil, err
+	}
+
+	return &Repository{
+		repo:     repo,
+		fallback: &gitcmd.Repository{Dir: dir},
+	}, nil
+}
+
+func (r *Repository) BlameFile(path string, opt *vcs.BlameOptions) ([]*vcs.Hunk, error) {
+	// TODO: Remove fallback usage: BlameFile
+	return r.fallback.BlameFile(path, opt)
+}
+
+func (r *Repository) Diff(base, head vcs.CommitID, opt *vcs.DiffOptions) (*vcs.Diff, error) {
+	// TODO: Remove fallback usage: Diff
+	return r.fallback.Diff(base, head, opt)
+}
+
+func (r *Repository) CrossRepoDiff(base vcs.CommitID, headRepo vcs.Repository, head vcs.CommitID, opt *vcs.DiffOptions) (*vcs.Diff, error) {
+	// TODO: Remove fallback usage: CrossRepoDiff
+	return r.fallback.CrossRepoDiff(base, headRepo, head, opt)
+}
+
+func (r *Repository) Search(at vcs.CommitID, opt vcs.SearchOptions) ([]*vcs.SearchResult, error) {
+	// TODO: Remove fallback usage: Search
+	return r.fallback.Search(at, opt)
+}
+
+func (r *Repository) MergeBase(a, b vcs.CommitID) (vcs.CommitID, error) {
+	// TODO: Remove fallback usage: MergeBase
+	return r.fallback.MergeBase(a, b)
+}
+
+// ResolveRevision returns the revision that the given revision
+// specifier resolves to, or a non-nil error if there is no such
+// revision.
+func (r *Repository) ResolveRevision(spec string) (vcs.CommitID, error) {
+	// TODO: git rev-parse supports a horde of complex syntaxes, it will be a fair bit more work to support all of them.
+	// e.g. "master@{yesterday}", "master~3", and various text/path/tree traversal search.
```

Comment:
Assuming vast majority of queries will be legit, it might be safe to just do a second fallback query whenever we fail to find something. Sound good?